### PR TITLE
add Player::getDeviceID,Player::getDeviceOS

### PR DESCRIPTION
**Pull request by Blackjack200**
## Introduction

update Player API
### API changes

add:
 - Player::getDeviceID
 - Player::getDeviceOS
### Behavioural changes

none
## Backwards compatibility

none

### CODE REVIEW

File: src/pocketmine/Player.php [hunk 1]
```diff
@@ -918,6 +918,27 @@ protected function checkItemCooldowns() : void{
...
+		return $this->deviceModel;
+	}
+
+	/**
+	 * @return mixed
+	 */
+	public function getDeviceId() {
+		return $this->deviceId;
+	}
+
```
Comment (TwistedAsylumMC) [inline src/pocketmine/Player.php:941]: You should remove the PHPDocs and add the correct type hints (not mixed)

File: src/pocketmine/Player.php [hunk 2]
```diff
@@ -1786,7 +1807,9 @@ public function checkNetwork(){
 			$this->batchedPackets = [];
 		}
 	}
-
+	private $deviceId;
+	private $deviceModel;
+	private $deviceOS;
```
Comment (TwistedAsylumMC) [inline src/pocketmine/Player.php:1812]: These are not in the right place, they should be before the constructor and they should have PHPDocs to documet their types

File: src/pocketmine/Player.php [hunk 3]
```diff
@@ -1840,17 +1863,17 @@ public function handleLogin(LoginPacket $packet) : bool{
...
 
 		if(count($this->server->getOnlinePlayers()) >= $this->server->getMaxPlayers() and $this->kick("disconnectionScreen.serverFull", false)){
 			return true;
 		}
-
 		$this->randomClientId = $packet->clientId;
-
+		$this->deviceId = $packet->clientData['DeviceId'];
+		$this->deviceModel = $packet->clientData['DeviceModel'];
+		$this->deviceOS = $packet->clientData['DeviceOS'];
```
Comment (TwistedAsylumMC) [inline src/pocketmine/Player.php:1876]: Double quotes should be used instead of single quotes

File: src/pocketmine/Player.php (hunk 1 above)
Comment (TwistedAsylumMC) [reply src/pocketmine/Player.php:941]: These methods are also in the wrong place, they are not near anything related to them.

File: src/pocketmine/Player.php (hunk 3 above)
Comment (Blackjack200) [reply src/pocketmine/Player.php:1876]: `$foo = [];
$bar = [];
for ($i = 0 ; $i <= 1000000 ; $i++) {
	$time1 = ms();
	echo "just a test" , PHP_EOL;
	$time2 = ms();
	$foo[] = $time2 - $time1;
	$time1 = ms();
	echo 'just a test' , PHP_EOL;
	$time2 = ms();
	$bar[] = $time2 - $time1;
}
arsort($foo);
arsort($bar);
echo '----- use double quotes time-----',PHP_EOL , '[minium]' , $foo[count($foo) - 1],PHP_EOL , '[maxium]' , $foo[0],PHP_EOL,'----- use single quotes time-----',PHP_EOL , '[minium]' , $bar[count($foo) - 1],PHP_EOL , '[maxium]' , $bar[0];
`
_----- use double quotes time-----
[minium]2.000000000002E-6
[maxium]3.9999999999984E-5
----- use single quotes time-----
[minium]1.9999999999742E-6
[maxium]2.9999999999752E-6_
i think using single quotes is faster than using double quotes,so why not use it?🤣

File: src/pocketmine/Player.php (hunk 2 above)
Comment (Blackjack200) [reply src/pocketmine/Player.php:1812]: `$foo = [];
$bar = [];
for ($i = 0 ; $i <= 1000000 ; $i++) {
	$time1 = ms();
	echo "just a test" , PHP_EOL;
	$time2 = ms();
	$foo[] = $time2 - $time1;
	$time1 = ms();
	echo 'just a test' , PHP_EOL;
	$time2 = ms();
	$bar[] = $time2 - $time1;
}
arsort($foo);
arsort($bar);
echo '----- use double quotes time-----',PHP_EOL , '[minium]' , $foo[count($foo) - 1],PHP_EOL , '[maxium]' , $foo[0],PHP_EOL,'----- use single quotes time-----',PHP_EOL , '[minium]' , $bar[count($foo) - 1],PHP_EOL , '[maxium]' , $bar[0];
`
_----- use double quotes time-----
[minium]2.000000000002E-6
[maxium]3.9999999999984E-5
----- use single quotes time-----
[minium]1.9999999999742E-6
[maxium]2.9999999999752E-6_
i think using single quotes is faster than using double quotes,so why not use it?🤣

Comment (Blackjack200) [reply src/pocketmine/Player.php:1812]: `function ms() {
	return explode(' ' , microtime())[0];
}`

Comment (TwistedAsylumMC) [reply src/pocketmine/Player.php]: Consistent code style

**Comment by Wumpotamus** [comment]
Should target next-minor branch, according to [semver standards](https://semver.org).

File: src/pocketmine/Player.php [hunk 4]
```diff
@@ -918,6 +923,8 @@ protected function checkItemCooldowns() : void{
 		}
 	}
 
+
+
```
Comment (dktapps) [inline src/pocketmine/Player.php:927]: unrelated formatting changes

File: src/pocketmine/Player.php [hunk 5]
```diff
@@ -1786,7 +1793,6 @@ public function checkNetwork(){
 			$this->batchedPackets = [];
 		}
 	}
-
```
Comment (dktapps) [inline src/pocketmine/Player.php:1789]: unrelated formatting change

File: src/pocketmine/Player.php [hunk 6]
```diff
@@ -1840,17 +1846,17 @@ public function handleLogin(LoginPacket $packet) : bool{
 		$this->username = TextFormat::clean($packet->username);
 		$this->displayName = $this->username;
 		$this->iusername = strtolower($this->username);
-
```
Comment (dktapps) [inline src/pocketmine/Player.php:1843]: and here

File: src/pocketmine/Player.php [hunk 7]
```diff
@@ -1840,17 +1846,17 @@ public function handleLogin(LoginPacket $packet) : bool{
...
 		if($packet->locale !== null){
 			$this->locale = $packet->locale;
 		}
 
 		if(count($this->server->getOnlinePlayers()) >= $this->server->getMaxPlayers() and $this->kick("disconnectionScreen.serverFull", false)){
 			return true;
 		}
-
```
Comment (dktapps) [inline src/pocketmine/Player.php:1851]: and here

File: src/pocketmine/Player.php [hunk 8]
```diff
@@ -3963,4 +3969,24 @@ public function getLoaderId() : int{
 	public function isLoaderActive() : bool{
 		return $this->isConnected();
 	}
+	/**
```
Comment (dktapps) [inline src/pocketmine/Player.php:3972]: missing newline above here

File: src/pocketmine/Player.php [hunk 9]
```diff
@@ -3963,4 +3969,24 @@ public function getLoaderId() : int{
...
+	 * @return int
+ 	*/
```
Comment (dktapps) [inline src/pocketmine/Player.php:3974]: improper docblock indentation; also this @return is useless

File: src/pocketmine/Player.php [hunk 10]
```diff
@@ -3963,4 +3969,24 @@ public function getLoaderId() : int{
...
+	public function getDeviceOS() : int{
+		return $this->deviceOS;
+	}
+
+	/**
+	 * @return string
```
Comment (dktapps) [inline src/pocketmine/Player.php:3980]: useless @return

File: src/pocketmine/Player.php [hunk 11]
```diff
@@ -3963,4 +3969,24 @@ public function getLoaderId() : int{
...
+	 */
+	public function getDeviceModel() : string {
```
Comment (dktapps) [inline src/pocketmine/Player.php:3982]: formatting inconsistent on return type

File: src/pocketmine/Player.php [hunk 12]
```diff
@@ -3963,4 +3969,24 @@ public function getLoaderId() : int{
...
+		return $this->deviceModel;
+	}
+
+	/**
+	 * @return string
```
Comment (dktapps) [inline src/pocketmine/Player.php:3987]: useless @return

File: src/pocketmine/Player.php [hunk 13]
```diff
@@ -3963,4 +3969,24 @@ public function getLoaderId() : int{
...
+	 */
+	public function getDeviceId() : string {
```
Comment (dktapps) [inline src/pocketmine/Player.php:3989]: return type formatting not consistent with the rest of the code

File: src/pocketmine/Player.php [hunk 14]
```diff
@@ -370,7 +370,12 @@ public static function isValidUserName(?string $name) : bool{
...
 	/** @var Vector3|null */
 	protected $lastRightClickPos = null;
-
+	/** @var string */
+	private $deviceId;
+	/** @var string */
+	private $deviceModel;
+	/** @var int */
+	private $deviceOS;
 	/**
```
Comment (dktapps) [inline src/pocketmine/Player.php:379]: not enough blank lines here
